Update print outputs in finetuning & pretraining

rasbt opened the pull request:
This is only for LoRA at the moment to show how it looks like and to discuss further before applying it to the other scripts:

```
Ep 1 | iter 1 | stEp 1: loss train (curr): 1.657, val (prev val iter): None | iter time: 298.27 ms
Ep 1 | iter 2 | stEp 1: loss train (curr): 1.506, val (prev val iter): None | iter time: 174.64 ms
Ep 1 | iter 3 | stEp 1: loss train (curr): 1.610, val (prev val iter): None | iter time: 106.92 ms
Ep 1 | iter 4 | step 1: loss train (curr): 1.585, val (prev val iter): None | iter time: 181.34 ms (opt.step)
Ep 1 | iter 5 | step 1: loss train (curr): 1.577, val (prev val iter): None | iter time: 154.41 ms
Ep 1 | iter 6 | step 1: loss train (curr): 1.737, val (prev val iter): None | iter time: 102.00 ms
Ep 1 | iter 7 | step 1: loss train (curr): 1.671, val (prev val iter): None | iter time: 156.28 ms
Ep 1 | iter 8 | step 2: loss train (curr): 1.761, val (prev val iter): None | iter time: 121.17 ms (opt.step)
Ep 1 | iter 9 | step 2: loss train (curr): 1.766, val (prev val iter): None | iter time: 151.38 ms
...
Ep 1 | iter 119 | step 29: loss train (curr): 1.199, val (prev val iter): None | iter time: 166.23 ms
Ep 1 | iter 120 | step 30: loss train (curr): 1.162, val (prev val iter): None | iter time: 164.51 ms (opt.step)
Validating ...
Recommend a movie for me to watch during the weekend and explain the reason.
Below is an instruction that describes a task. Write a response that appropriately completes the request.

### Instruction:
Recommend a movie for me to watch during the weekend and explain the reason.

### Response:
Hey! I just watched the movie, "The Martian" on Netflix. It was a good movie and I enjoy watching movies on Netflix because it offers me a ton of different movies to watch. The movie, "The Martian" is about an astronaut in a space mission that was sent to explore Mars. I recommend watching the movie, "The Martian" because it is a worthwhile movie that has a good story, characters, and a good
iter 120: val loss 1.2820, val time: 8758.00 ms
Ep 1 | iter 121 | step 30: loss train (curr): 1.190, val (prev val iter): 1.282 | iter time: 114.93 ms
Ep 1 | iter 122 | step 30: loss train (curr): 1.232, val (prev val iter): 1.282 | iter time: 96.62 ms
Ep 1 | iter 123 | step 30: loss train (curr): 1.240, val (prev val iter): 1.282 | iter time: 153.89 ms
Ep 1 | iter 124 | step 31: loss train (curr): 1.271, val (prev val iter): 1.282 | iter time: 167.93 ms (opt.step)
...
Ep 1 | iter 238 | step 59: loss train (curr): 1.129, val (prev val iter): 1.282 | iter time: 149.74 ms
Ep 1 | iter 239 | step 59: loss train (curr): 1.122, val (prev val iter): 1.282 | iter time: 148.59 ms
Ep 1 | iter 240 | step 60: loss train (curr): 1.139, val (prev val iter): 1.282 | iter time: 167.52 ms (opt.step)
Validating ...
Recommend a movie for me to watch during the weekend and explain the reason.
Below is an instruction that describes a task. Write a response that appropriately completes the request.

### Instruction:
Recommend a movie for me to watch during the weekend and explain the reason.

### Response:
Dear Sir,

The holiday season is fast approaching, and you have to plan a trip for the family, which brings you to the necessity of choosing a good movie for you and your family.

But I'm afraid I cannot recommend you a good movie just now. However, I can suggest you a movie that will suit your needs.

First of all, this movie is in a 3D digital format, which makes watching it pleasurable. Moreover,
iter 240: val loss 1.1173, val time: 8478.09 ms
Ep 1 | iter 241 | step 60: loss train (curr): 1.141, val (prev val iter): 1.117 | iter time: 123.90 ms
Ep 1 | iter 242 | step 60: loss train (curr): 1.164, val (prev val iter): 1.117 | iter time: 95.84 ms
Ep 1 | iter 243 | step 60: loss train (curr): 1.133, val (prev val iter): 1.117 | iter time: 164.39 ms
Ep 1 | iter 244 | step 61: loss train (curr): 1.137, val (prev val iter): 1.117 | iter time: 133.60 ms (opt.step)
Ep 1 | iter 245 | step 61: loss train (curr): 1.215, val (prev val iter): 1.117 | iter time: 172.17 ms
```

What's new are 

1. the `Ep` number
2.  the `val` loss
3.  `optimizer.step` is shortened to `opt.step`

I feel like it's still too long. Regarding the validation loss, the `prev val iter` in `val (prev val iter)` should indicate that it is not calculated fresh in this step. Maybe we can drop it and assume that people will know this? In the PyTorch Lightning Trainer, we do something similar where we only update the validation loss after each epoch or so by default but still display the value in the progress bar.


Fixes #1042 and #1015

awaelchli commented on litgpt/finetune/lora.py:
```diff
@@ -262,10 +263,15 @@ def fit(
-                f"iter {metrics['iter']} | step {metrics['step']}: loss {metrics['loss']:.4f}, iter time:"
-                f" {metrics['iter_time'] * 1000:.2f} ms{' (optimizer.step)' if not is_accumulating else ''}"
+                f"Ep {metrics['epoch']+1} | iter {metrics['iter']} | step {metrics['step']}:"
+                f" loss train (curr): {metrics['loss']:.3f},"
+                f" val (prev val iter): {val_loss_str} |"
```
My preference is to keep the information compact.
1) I don't think anyone is questioning whether the training loss is from any other iteration than the current one, especially when it is numbered by the iteration in the same line. So I suggest to drop (curr).
2) Same argument for val, why wouldn't it be the value from the previous validation? And we don't validate every training iteration.

awaelchli commented on litgpt/finetune/lora.py:
```diff
@@ -262,10 +263,15 @@ def fit(
             fabric.print(
-                f"iter {metrics['iter']} | step {metrics['step']}: loss {metrics['loss']:.4f}, iter time:"
-                f" {metrics['iter_time'] * 1000:.2f} ms{' (optimizer.step)' if not is_accumulating else ''}"
+                f"Ep {metrics['epoch']+1} | iter {metrics['iter']} | step {metrics['step']}:"
```
Maybe the full word Epoch is better (same in Lightning Trainer).
I agree with shortening optimizer.step

rasbt commented:
Thanks for the feedback. That sounds great, so we can greatly simplify. I updated the scripts so that the output is now as follows:

```
Epoch 1 | iter 59 | step 1: loss train: 2.182, val: n/a | iter time: 144.97 ms
Epoch 1 | iter 60 | step 1: loss train: 2.180, val: n/a | iter time: 220.11 ms
Epoch 1 | iter 61 | step 1: loss train: 2.191, val: n/a | iter time: 125.79 ms
Epoch 1 | iter 62 | step 1: loss train: 2.192, val: n/a | iter time: 149.53 ms
Epoch 1 | iter 63 | step 1: loss train: 2.182, val: n/a | iter time: 188.16 ms
Epoch 1 | iter 64 | step 2: loss train: 2.175, val: n/a | iter time: 144.90 ms (step)
Validating ...
Recommend a movie for me to watch during the weekend and explain the reason.
Below is an instruction that describes a task. Write a response that appropriately completes the request.

### Instruction:
Recommend a movie for me to watch during the weekend and explain the reason.

### Response:
As a MovieRecommender, I will construct my recommendation by leveraging the following factors:

- Runtime
- RecommenderSystem type
- Genre
- Platform(TV,Web,Mobile)
- Award(Oscar,Grammy)

An example of recommendation for movie genre is as follows:

- If movie genre is action, I will select movie which is action movie with runtime(Trailer video is good for recommendation)
- The movie
iter 64: val loss 2.1240, val time: 11392.31 ms
Epoch 1 | iter 65 | step 2: loss train: 2.166, val: 2.124 | iter time: 126.13 ms
Epoch 1 | iter 66 | step 2: loss train: 2.166, val: 2.124 | iter time: 103.34 ms
Epoch 1 | iter 67 | step 2: loss train: 2.156, val: 2.124 | iter time: 186.89 ms
Epoch 1 | iter 68 | step 2: loss train: 2.173, val: 2.124 | iter time: 113.88 ms
...
Epoch 1 | iter 92 | step 2: loss train: 2.117, val: 2.124 | iter time: 127.20 ms
Epoch 1 | iter 93 | step 2: loss train: 2.118, val: 2.124 | iter time: 101.81 ms
Epoch 1 | iter 94 | step 2: loss train: 2.113, val: 2.124 | iter time: 226.40 ms
Epoch 1 | iter 95 | step 2: loss train: 2.121, val: 2.124 | iter time: 182.90 ms
Epoch 1 | iter 96 | step 3: loss train: 2.124, val: 2.124 | iter time: 214.36 ms (step)
Epoch 1 | iter 97 | step 3: loss train: 2.137, val: 2.124 | iter time: 166.88 ms
```

awaelchli reviewed (approved):
Could you also do similar changes to pretrain.py?

Maybe the little vertical bar `|` between iter and step could move one column to the right to do the grouping of these things

carmocca reviewed (approved):
I prefer not including the val loss on every train print BUT since I'm outnumbered let's do it

rasbt commented:
Updated the print output organization as suggested by @awaelchli : 

```
Epoch 1 | iter 61 step 0 | loss train: 2.294, val: n/a | iter time: 90.73 ms
Epoch 1 | iter 62 step 0 | loss train: 2.287, val: n/a | iter time: 91.05 ms
Epoch 1 | iter 63 step 0 | loss train: 2.289, val: n/a | iter time: 91.16 ms
Epoch 1 | iter 64 step 1 | loss train: 2.296, val: n/a | iter time: 300.39 ms (step)
Epoch 1 | iter 65 step 1 | loss train: 2.305, val: n/a | iter time: 65.25 ms
Epoch 1 | iter 66 step 1 | loss train: 2.308, val: n/a | iter time: 90.13 ms
Epoch 1 | iter 67 step 1 | loss train: 2.307, val: n/a | iter time: 91.03 ms
Epoch 1 | iter 68 step 1 | loss train: 2.303, val: n/a | iter time: 89.79 ms
...
```

I also updated the pretraining script. However, trying to test the pretraining scripts, I am getting weird errors with the data prep:

```
python litgpt/data/prepare_starcoder.py \
  --input_dir data/starcoderdata-raw \
  --output_dir data/starcoder \
  --tokenizer_path checkpoints/meta-llama/Llama-2-7b-hf
```

```
/commands/python: line 37: 164460 Segmentation fault      (core dumped) "$
```

and

```
python litgpt/data/prepare_slimpajama.py \
  --input_dir data/slimpajama-raw/validation \
  --output_dir data/slimpajama/val \
  --tokenizer_path checkpoints/meta-llama/Llama-2-7b-hf
```

```
Traceback (most recent call last):
  File "litgpt/data/prepare_slimpajama.py", line 3, in <module>
    import json
  File "/teamspace/studios/this_studio/lit-gpt/litgpt/data/json.py", line 8, in <module>
    import torch
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.8/site-packages/torch/__init__.py", line 1215, in <module>
    from .storage import _StorageBase, TypedStorage, _LegacyStorage, UntypedStorage, _warn_typed_storage_removal
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.8/site-packages/torch/storage.py", line 14, in <module>
    import numpy as np
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.8/site-packages/numpy/__init__.py", line 141, in <module>
    from . import core
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.8/site-packages/numpy/core/__init__.py", line 9, in <module>
    from numpy.version import version as __version__
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.8/site-packages/numpy/version.py", line 7, in <module>
    vinfo: dict[str, str] = get_versions()
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.8/site-packages/numpy/_version.py", line 21, in get_versions
    return json.loads(version_json)
AttributeError: partially initialized module 'json' has no attribute 'loads' (most likely due to a circular import)
```

Have you seen this before?

carmocca commented:
We should rename https://github.com/Lightning-AI/litgpt/blob/wip/litgpt/data/json.py so that it doesn't collide with the standard library module

rasbt commented:
I just renamed it to `json_utils` here, but I can also do that in a separate PR if that's preferred.

carmocca commented on litgpt/pretrain.py:
```diff
@@ -251,11 +252,15 @@ def fit(
-                f" {metrics['iter_time'] * 1000:.2f} ms{' (optimizer.step),' if not is_accumulating else ','}"
+                f"Epoch {metrics['epoch']+1} | iter {metrics['iter']} step {metrics['step']} |"
+                f" loss train: {metrics['loss']:.3f},"
+                f" val: {val_loss_str} |"
```
You'll need to define the variable

carmocca reviewed (commented):
The failing tests need to be updated. They assert the stdout of some of the scripts, particularly the number of `optimizer.step()` prints

awaelchli commented on litgpt/data/__init__.py:
```diff
@@ -4,7 +4,7 @@
 from litgpt.data.alpaca_2k import Alpaca2k
 from litgpt.data.alpaca_gpt4 import AlpacaGPT4
-from litgpt.data.json import JSON
+from litgpt.data.json_utils import JSON
```
I find utils not suitable here because we already have a utils.py for other purposes. How about `json_data`?

rasbt replied:
That's a good idea. Renamed it!

awaelchli commented on litgpt/pretrain.py:
```diff
@@ -251,11 +252,16 @@ def fit(
                     state["iter_num"] * train.micro_batch_size * model.max_seq_length * fabric.world_size
                 ),
                 "learning_rate": lr,
+                "val_loss": val_loss,
```
We can't log the loss here: It is None in the first iteration which is annoying to handle, and we are logging it below after the val loop. 
```suggestion
```

carmocca commented on litgpt/finetune/adapter.py:
```diff
@@ -209,9 +210,13 @@ def fit(
                 time=t1 - total_t0, batches=iter_num, samples=iter_num * train.micro_batch_size, lengths=total_lengths
             )
             throughput.compute_and_log(step=iter_num)
+            val_loss_str = 'n/a' if val_loss is None else f"{val_loss:.3f}"
```
Should we cache the `val_loss` string? Validation doesn't run often so we could optimize this a bit by reusing the string that we already computed.

`val_loss` could be initialized to `n/a` and then we only recompute it here if it's a tensor

```suggestion
            if isinstance(val_loss, torch.Tensor):
                val_loss = f"{val_loss:.3f}"
```

carmocca replied:
I profiled 3 prints of the same cuda tensor. PyTorch seems to be smart already and the time it takes for the second and third prints is much lower (see highlighted section compared to the left side). But it still triggers stream synchronization that we could simply avoid

![image](https://github.com/Lightning-AI/litgpt/assets/11067892/3a0450e6-e16a-441f-878f-1e9808198fc7)

rasbt replied:
Wow nice. Yeah, let's use your suggestion. I added it to all the scripts. Note that in the pretraining script, the `val_loss` is a `float` not a `torch.Tensor`.

carmocca replied:
Are you sure about that? The validaiton function is annotated as returning a Tensor

carmocca replied:
Ah thanks. I missed that. In that case my suggestion is not relevant for pretrain.py. Looks good anyways
